### DISCUSSION TITLE
[WIDGET] Trilium Hexmap - A widget displaying fetched notes as places and people on top of a hex grid map

### DISCUSSION THREAD

**Discussion by MorganeLeCaignec** [TriliumNext/Trilium#1639] [Show and tell]
UPDATE: this is still regularly updated and has been cleaned-up and made more easily configurable, check the repo and the README of that repo, link below.

https://gitlab.com/MorganeLeCaignec/trilium-hexmap

---

Hello there,

In my quest to automate things and keep track of stuff for my tabletop campaign (the main reason I use Trilium!), i've made this frankly ridiculous widget that generates a hex grid (based on promoted attributes for max x/y size, start/end values of the range of coordinates to display, and hex size which is basically a zoom level), fetches Location notes with specific attributes (in my case hex coordinates, and an optional icon to be displayed on the hex map) and displays them at their proper coordinates with conditional colors (is the note TODO, is it empty, etc) and a clickable link.
It has an optional background image url which will appear behind the grid.
It even fetches notes tagged as "agents" (NPCs) and if they are related to a location note it changes the hex's background color, the color depending on the Agent type (neutral, hostile, allied, etc..).

This seemed overkill at first but honestly it's incredibly useful, I can write and update my notes and just with a tiny bit of setting up for labels, it all shows up on this map with everything clear and easy to track!

I've used it in-game already, tracking the player's movement and clicking on the link for the prepped hex note :)
It's also convenient for me to be able to quickly tell what i've not prepared and what I need to fill in.

Had to figure out some math for the hex grid itself but that wasn't too hard, works super well.

![image](https://user-images.githubusercontent.com/12102823/138761049-802dc78f-c628-4841-bfdf-f133989ae81d.png)

**Comment by DannyFL** [comment]
@QuentinLeCaignec 
So I may be stupid... and while I'm not okay with that, I can acknowledge it from time to time. That said, how exactly does this get integrated? I attempted to do a basic import and implement. After banging my head against the wall for an hour or so it still didn't work at all for me. Would you be willing to write a more detailed step by step implementation for me?

Cheers

**Reply by MorganeLeCaignec** [reply]
Hey, I didn't think anyone had actually tried!
So I think the import/export stuff doesn't even work if you just download the repo and try to import it, and I'm not sure why.
I think the import/export system was changed at some point between 0.45 and 0.47, or maybe Trilium just doesn't want to import a modified .zip, because even if I just keep the files from the original exported .zip intact and put those in the repo source, re-zipping them and importing them does not work.

So instead I've updated it and just put the actual .zip into the repo. It's the exported .zip that Trilium gave me unmodified, and now it works. I've added installation instructions and usage instructions to the readme too.

I forgot about this and I thought nobody had tried, but if this can actually be useful to others I might go back to my mess of a code and try to improve it. Maybe add support for line-aligned maps (currently only support column-aligned maps because that's what I need), or add some editable margin + offset attributes to make the lining up with your background image less of a nightmare.
As I've said in the README, right now it works for my own map which assumes a column-aligned map with no margin at all and very thin lines on the grid. But that's only if you want to line it up with a preexisting grid.

I'd love to know if you can get it to work, and if you have any ideas for improvements on attributes, UI or features. I'm a bit stumped on the features part because it's hard to put a lot of readable data inside of a hex, but I'm sure I can figure stuff out if I find the motivation.

**Reply by DannyFL** [reply]
Alright, so I was able to import it successfully and pull up your pre-rendered map. I found the only reference to an image was in an owned attr on the widget page. I uploaded my own image to imgur, and replaced the link. Now all I am looking at is a black screen full of mini notes. 
![Capture](https://user-images.githubusercontent.com/35320074/128617809-b9417f2d-94c6-4b6d-b3ea-267a3f8362ce.PNG)
Not sure if I'm missing something here. When I attempted to re-render it, it gave me an error pointing to the render.html. When I open up the render.html, the first line is highlighted in red: <div class="hex-container" id="container"></div> 

That's where it's currently sitting, I'm continuing to dig around a little bit more, but wanted to know if I'm missing something here.

**Reply by DannyFL** [reply]
Thought: Could you try to duplicate it with my image and see if it works for you? https://imgur.com/gallery/2fhA8Pt is the image I want to use. Would probably be easier from a user pov to be able to just have a .jpg in the folder, which can be rendered that way instead of using an external link -- might speed things up just a hair.

**Reply by MorganeLeCaignec** [reply]
Did you redownload the widget from the repo? It looks like the image background link you have is the old one which yes is a private link. the image background should be "https://i.imgur.com/XNMEKeI.jpeg" and not the one you have.

The widget was updated a bit so it would be better to use the latest version that's on the repo right now.

I also tried with your image, and it works. But you have to be careful, the link you sent me is a gallery link, the actual image itself. You need to right click on your image on the imgur page and open the actual image, and copy that link (in your case that's "https://i.imgur.com/ypKcAq1.jpeg").

Here's a screenshot of it with your image, 
![image](https://user-images.githubusercontent.com/12102823/128630571-ddc3c381-1a24-4300-90bb-c3e2c5a17f1f.png)

I aligned the number of x/y with the image size, this is something I should probably improve on and find an actual easy method, calculating a divisible number for width/height of the image, but it's weird for hex grids because the X doesn't like the Y, since the Y has those shifting lines... anyway, something to figure out and make instructions for so it's less annoying. Right now I just put some estimated numbers and adjusted by raising/lowering the X or Y until it matched the right and bottom edges. You could also multiply those X/Y values (15 and 23) to 30 and 46, etc.. to change the scale of hexes.

**Reply by MorganeLeCaignec** [reply]
I am going to try and improve things today, see if I can't make it work better in general and be more intuitive, as well as fix a few bugs that seem to have appeared since 0.45 (like clicking an empty hex label, it should normally create a new empty note at those coordinates but that doesn't work anymore).

**Reply by DannyFL** [reply]
Awesome, it did work with that change to the URL right off the bat, I'm feeling stupid for messing that up. Yeah, I did download the new one with the updated readme, etc. Just let me know when to try the updated one and I'll give that a shot. Thank you for bearing with me!

**Reply by DannyFL** [reply]
Quick update:

I am able to manually add values to respective hexes, which works fine. Still getting the error on "click-adding" -- not a huge deal, but would definitely be more convenient. Anyway, thank you so much for doing this! It's really going to pay off once I finish plugging my own content into it.

Thank you again @QuentinLeCaignec

**Reply by MorganeLeCaignec** [reply]
Hey, glad you find it useful!

I just pushed a new version to the repo, with some fixes and improvements.

You can now reference an Image Note to be used as the background image through a relation attribute, you don't need to reference an external URL although that's still an option!
I fixed up some errors and weirdness in the code. I added a bunch of working examples of notes, a template and how to make note creation work (you can double-click the label of an empty hex to create a new note if you properly set-up the attributes, the new version now has a working example with explanations). It still throws an error on creation but that seems to be a problem from the backend API of trilium, and I made an Issue about to see if that can be fixed.

my next step is probably going to be a polish on the code and making the attributes clearer (better names, etc), and trying to add a basic in-window UI that could do some stuff better like zooming in/out or toggling information on/off without having to modify Promoted Attributes which is a bit cumbersome.

**Reply by DannyFL** [reply]
That's awesome, thank you so much for all the time you have put into this! I was really hoping not to have to port my D&D notes over to OneNote for this purpose. Do let me know when you have a zoom in/out solution working, and I'll be happy to be a second set of eyes/tester for this. Cheers!

**Reply by MorganeLeCaignec** [reply]
Alright!
New version (1.0) with a bunch of stuff.
I moved stuff around and polished a few things. I changed the attributes to more intuitive names as well as removed the ones that don't need to be "promoted" so as to not clutter that tab. The ranges are now only two attributes (with two values separated with a comma) instead of 4, the "hex_height" attribute is called scale and is now controllable by clicking on the +/- buttons at the bottom right (changing the scale redraws the widget every time unfortunately, no way around that so far).
There's two ways of adding a background map, by linking to an image note or with an external URL.
Automatic creation of notes by double-clicking the label of an empty hex works, I removed the annoying errors that popped up when creating a note. You can also specify another note under which you want the new notes to be created, and you can specify a template note which will be used when creating a new notes (the working example has a template with pre-created hex_x and hex_y promoted attributes which is very useful).
I also changed the way the agents display on a hex, it can now handle multiple agents present on the same hex and displays each as little marker at the top of the hex, colored depending on the agent type. Those markers can be hovered and clicked which is way more convenient to keep track of who's where without having to open the location note first. If there are more than 4 agents on the same hex, a little "+" marker appears to indicate that some are now shown.

The new version also has a bunch more example notes all working to better get started on using it.

Thanks for your input and for motivating me to improve this :)

**Reply by DannyFL** [reply]
One thing I noticed was that on the promoted attributes tab, changing those values did not actually change the scale of the map. Not actually sure what it did, I wound up having to change it under owned attributes and that took effect. The only promoted attr that made a difference was the scale, which effectively turns into the zoom function. If that is how it is supposed to be, that's cool, if not, sorry to be that guy. lol, I'm happy with it the way it is now, so feel free to not update it until you feel like it. Thanks again for the help with all this!

Cheers!

**Reply by MorganeLeCaignec** [reply]
Yes, the scale is the zoom level, controllable either by changing the value (and re-rendering the widget, as always) or with the +/- buttons. The other values are ranges to render only specific part of the full map, which is useful to me because my map is enormous (80 by 60) and I'll enter for example "20,40" X and "20,30" Y in order to only render a smaller rectangle of the map, which means I can have the automatic centering of the view where I want and also improve performance because my huge map can be a bit slow when fully rendered. Keep in mind any change to attributes on a render note needs a re-render (top-right, Actions, re-render).

I updated the instructions on the README of the code repo if you missed it, explaining the attributes and what they do.

I'm glad you're happy with it. I might have other ideas for it but for my own needs it's very good right now. There's a few adjustments (some CSS tweaking, some error-proofing, default values when some attributes are not set, etc..) that's I've added and I'll push those in a new version at some point. I probably also need to fix my Timeline widget, i'm assuming that repo is broken like this one was.

For reference, this is what my own setup looks like, and it is legitimately very useful to keep track of locations, factions and agents running around my sandbox. If you need an idea of what it looks like with a lot of content : 
![image](https://user-images.githubusercontent.com/12102823/128791037-80da06a2-34e3-41be-b4e7-77e10363c8f0.png)

**Comment by MorganeLeCaignec** [comment]
This will probably be the last update for now, but here's a new version, 1.2 : 

I added toggle attributes to shift the map layout to vertical/horizontal, and to odd/even shifting, which means this should now work with any type of hex grid! This took some work, but as far as I've tested it seems like it works without issue.

I polished a few things like adding attribute declarations (even the non-promoted ones, more convenient to promote them if you need it), added a few "location types" icons in the CSS, added a black border to agent markers, tweaked some more CSS, updated the README.

**Reply by MorganeLeCaignec** [reply]
Small update 1.3 with a few CSS fixes (agent markers glitching on hover or cut-off at the hex border, adjusted size)

**Comment by MorganeLeCaignec** [comment]
Updated to 1.4 for trilium version 0.48.3, fixes the API breaking changes.
Changed the hex note detection to use only one attribute called "hex_coords" instead of two called "hex_x" and "hex_y", the "hex_coords" attribute is in the X.Y format where X and Y (separated by a dot) at the coordinates of the hex.

**Comment by MorganeLeCaignec** [comment]
Updated to 1.5 for trilium 0.54.2, refactors/cleans up code, fixes some breaking changes for view centering, zoom buttons placement/working, and making the render drag-scrollable.

**Comment by MorganeLeCaignec** [comment]
Updated to 1.7 for 0.62.4 !
